# SO IN THIS NOTEBOOK WE'RE GIVEN ZOMATO DATASET ON BANGLORE CITY WHICH IS COMPOSED OF MANY FEATURES AND WE'RE TO PREDICT THE RATINGS GIVEN BY THE CUSTOMERS TO VARIOUS RESTRAUNTS IN THE CITY ON THE ZOMATO APP. SO HERE WE GO......

In [ ]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
df = pd.read_csv('zomato.csv')

In [4]:
df['listed_in(city)'].unique()

array(['Banashankari', 'Bannerghatta Road', 'Basavanagudi', 'Bellandur',
       'Brigade Road', 'Brookefield', 'BTM', 'Church Street',
       'Electronic City', 'Frazer Town', 'HSR', 'Indiranagar',
       'Jayanagar', 'JP Nagar', 'Kalyan Nagar', 'Kammanahalli',
       'Koramangala 4th Block', 'Koramangala 5th Block',
       'Koramangala 6th Block', 'Koramangala 7th Block', 'Lavelle Road',
       'Malleshwaram', 'Marathahalli', 'MG Road', 'New BEL Road',
       'Old Airport Road', 'Rajajinagar', 'Residency Road',
       'Sarjapur Road', 'Whitefield'], dtype=object)

In [5]:
df.shape

(51717, 17)

In [6]:
df

url  \
0      https://www.zomato.com/bangalore/jalsa-banasha...   
1      https://www.zomato.com/bangalore/spice-elephan...   
2      https://www.zomato.com/SanchurroBangalore?cont...   
3      https://www.zomato.com/bangalore/addhuri-udupi...   
4      https://www.zomato.com/bangalore/grand-village...   
...                                                  ...   
51712  https://www.zomato.com/bangalore/best-brews-fo...   
51713  https://www.zomato.com/bangalore/vinod-bar-and...   
51714  https://www.zomato.com/bangalore/plunge-sherat...   
51715  https://www.zomato.com/bangalore/chime-sherato...   
51716  https://www.zomato.com/bangalore/the-nest-the-...   

                                                 address  \
0      942, 21st Main Road, 2nd Stage, Banashankari, ...   
1      2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...   
2      1112, Next to KIMS Medical College, 17th Cross...   
3      1st Floor, Annakuteera, 3rd Stage, Banashankar...   
4      10, 3rd Floor, Lakshmi Associates, Gandhi Baza...   
...                                                  ...   
51712  Four Points by Sheraton Bengaluru, 43/3, White...   
51713  Number 10, Garudachar Palya, Mahadevapura, Whi...   
51714  Sheraton Grand Bengaluru Whitefield Hotel & Co...   
51715  Sheraton Grand Bengaluru Whitefield Hotel & Co...   
51716  ITPL Main Road, KIADB Export Promotion Industr...   

                                                    name online_order  \
0                                                  Jalsa          Yes   
1                                         Spice Elephant          Yes   
2                                        San Churro Cafe          Yes   
3                                  Addhuri Udupi Bhojana           No   
4                                          Grand Village           No   
...                                                  ...          ...   
51712  Best Brews - Four Points by Sheraton Bengaluru...           No   
51713                           Vinod Bar And Restaurant           No   
51714  Plunge - Sheraton Grand Bengaluru Whitefield H...           No   
51715  Chime - Sheraton Grand Bengaluru Whitefield Ho...           No   
51716                       The Nest - The Den Bengaluru           No   

      book_table    rate  votes                             phone  \
0            Yes   4.1/5    775    080 42297555\r\n+91 9743772233   
1             No   4.1/5    787                      080 41714161   
2             No   3.8/5    918                    +91 9663487993   
3             No   3.7/5     88                    +91 9620009302   
4             No   3.8/5    166  +91 8026612447\r\n+91 9901210005   
...          ...     ...    ...                               ...   
51712         No  3.6 /5     27                      080 40301477   
51713         No     NaN      0                    +91 8197675843   
51714         No     NaN      0                               NaN   
51715        Yes  4.3 /5    236                      080 49652769   
51716         No  3.4 /5     13                    +91 8071117272   

                         location            rest_type  \
0                    Banashankari        Casual Dining   
1                    Banashankari        Casual Dining   
2                    Banashankari  Cafe, Casual Dining   
3                    Banashankari          Quick Bites   
4                    Basavanagudi        Casual Dining   
...                           ...                  ...   
51712                  Whitefield                  Bar   
51713                  Whitefield                  Bar   
51714                  Whitefield                  Bar   
51715  ITPL Main Road, Whitefield                  Bar   
51716  ITPL Main Road, Whitefield   Bar, Casual Dining   

                                              dish_liked  \
0      Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1      Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2      Churros, Cannelloni, 

In [7]:
df['approx_cost(for two people)'].isnull().sum()

346

In [8]:
df['approx_cost(for two people)'].unique()

array(['800', '300', '600', '700', '550', '500', '450', '650', '400',
       '900', '200', '750', '150', '850', '100', '1,200', '350', '250',
       '950', '1,000', '1,500', '1,300', '199', '80', '1,100', '160',
       '1,600', '230', '130', '50', '190', '1,700', nan, '1,400', '180',
       '1,350', '2,200', '2,000', '1,800', '1,900', '330', '2,500',
       '2,100', '3,000', '2,800', '3,400', '40', '1,250', '3,500',
       '4,000', '2,400', '2,600', '120', '1,450', '469', '70', '3,200',
       '60', '560', '240', '360', '6,000', '1,050', '2,300', '4,100',
       '5,000', '3,700', '1,650', '2,700', '4,500', '140'], dtype=object)

In [138]:
df['approx_cost(for two people)'][0]

800.0

### SO, HERE WE BUILD A FUNCTION WHICH WE WILL CONVERT THE 'approx_cost(for two people )' COLUMN TO A PURE INTEGER FOR THE MODEL TO FEED ON IT

In [10]:
def remove(s):
    if(isinstance(s, str)):
        ans=0
        for i in s:
            if(i!=','):
                ans*=10
                ans+=int(i)
        return int(ans)
    else:
        return s
        

    

In [11]:
for i in range (len(df['approx_cost(for two people)'])):
    df.at[i,'approx_cost(for two people)']=remove(df.at[i,'approx_cost(for two people)'])
    

In [12]:
df['approx_cost(for two people)'].median()

400.0

In [13]:
df['approx_cost(for two people)'].fillna(400.0,inplace = True)

In [14]:
df['approx_cost(for two people)'].isnull().sum()

0

In [15]:
df['online_order'].unique()

array(['Yes', 'No'], dtype=object)

In [139]:
df['online_order'] = df['online_order'].replace({'Yes':1,'No':0})

/var/folders/jm/fzpx9nmx3gl9k9bfkhp1cy100000gn/T/ipykernel_21032/1190499278.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['online_order'] = df['online_order'].replace({'Yes':1,'No':0})


### AS THE 'online_order' COLUMN HAVE ONLY TWO UNIQUE VALUES WHICH ARE 'Yes' AND 'No' ,  WE'LL CONVERT THEM TO '1' AND '0' RESPECTIVELY SO THAT MODEL CAN FEED ON IT

In [17]:
df['online_order']

0        1
1        1
2        1
3        0
4        0
        ..
51712    0
51713    0
51714    0
51715    0
51716    0
Name: online_order, Length: 51717, dtype: int64

In [140]:
df['book_table'].unique()

array([1, 0])

### WE'LL CONVERT THE 'book_table' COLUMN AS WE DID WITH 'online_order' COLUMN

In [19]:
df['book_table'] = df['book_table'].replace({'Yes':1,'No':0})

In [20]:
df['book_table'].unique()

array([1, 0])

In [21]:
df['votes'].isnull().sum()

0

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 17 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   url                          51717 non-null  object 
 1   address                      51717 non-null  object 
 2   name                         51717 non-null  object 
 3   online_order                 51717 non-null  int64  
 4   book_table                   51717 non-null  int64  
 5   rate                         43942 non-null  object 
 6   votes                        51717 non-null  int64  
 7   phone                        50509 non-null  object 
 8   location                     51696 non-null  object 
 9   rest_type                    51490 non-null  object 
 10  dish_liked                   23639 non-null  object 
 11  cuisines                     51672 non-null  object 
 12  approx_cost(for two people)  51717 non-null  float64
 13  reviews_list    

In [23]:
df['listed_in(type)'].isnull().sum()

0

In [24]:
df['listed_in(type)'].unique()

array(['Buffet', 'Cafes', 'Delivery', 'Desserts', 'Dine-out',
       'Drinks & nightlife', 'Pubs and bars'], dtype=object)

In [141]:
df.corr()

online_order  book_table      rate     votes  \
online_order                     1.000000   -0.129724 -0.036798 -0.078384   
book_table                      -0.129724    1.000000  0.353517  0.324970   
rate                            -0.036798    0.353517  1.000000  0.380791   
votes                           -0.078384    0.324970  0.380791  1.000000   
location                        -0.096029    0.109832  0.110220  0.043839   
rest_type                        0.053374   -0.215334 -0.148186 -0.091944   
cuisines                        -0.019477    0.006974 -0.103791 -0.049121   
approx_cost(for two people)     -0.286137    0.604599  0.312006  0.310087   
menu_item                       -0.345250    0.083767  0.005221  0.030030   

                             location  rest_type  cuisines  \
online_order                -0.096029   0.053374 -0.019477   
book_table                   0.109832  -0.215334  0.006974   
rate                         0.110220  -0.148186 -0.103791   
votes                        0.043839  -0.091944 -0.049121   
location                     1.000000  -0.077308 -0.001734   
rest_type                   -0.077308   1.000000  0.257848   
cuisines                    -0.001734   0.257848  1.000000   
approx_cost(for two people)  0.200471  -0.253287 -0.047604   
menu_item                    0.018427  -0.037843  0.032915   

                             approx_cost(for two people)  menu_item  
online_order                                   -0.286137  -0.345250  
book_table                                      0.604599   0.083767  
rate                                            0.312006   0.005221  
votes                                           0.310087   0.030030  
location                                        0.200471   0.018427  
rest_type                                      -0.253287  -0.037843  
cuisines                                       -0.047604   0.032915  
approx_cost(for two people)                     1.000000   0.127298  
menu_item                                       0.127298   1.000000

### HERE WE'RE DROPPING SOME COLUMNS ('phone',address','url') AS THEY ARE NOT AS SIGNIFICANT AND WON'T MAKE MUCH OF A DIFFERENCE ON PREDICTABILITY

In [26]:
df = df.drop(["phone","address","url","name"],axis = 1)  

In [27]:
df

online_order  book_table    rate  votes                    location  \
0                 1           1   4.1/5    775                Banashankari   
1                 1           0   4.1/5    787                Banashankari   
2                 1           0   3.8/5    918                Banashankari   
3                 0           0   3.7/5     88                Banashankari   
4                 0           0   3.8/5    166                Basavanagudi   
...             ...         ...     ...    ...                         ...   
51712             0           0  3.6 /5     27                  Whitefield   
51713             0           0     NaN      0                  Whitefield   
51714             0           0     NaN      0                  Whitefield   
51715             0           1  4.3 /5    236  ITPL Main Road, Whitefield   
51716             0           0  3.4 /5     13  ITPL Main Road, Whitefield   

                 rest_type                                         dish_liked  \
0            Casual Dining  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1            Casual Dining  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2      Cafe, Casual Dining  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3              Quick Bites                                        Masala Dosa   
4            Casual Dining                                Panipuri, Gol Gappe   
...                    ...                                                ...   
51712                  Bar                                                NaN   
51713                  Bar                                                NaN   
51714                  Bar                                                NaN   
51715                  Bar                       Cocktails, Pizza, Buttermilk   
51716   Bar, Casual Dining                                                NaN   

                                     cuisines  approx_cost(for two people)  \
0              North Indian, Mughlai, Chinese                        800.0   
1                 Chinese, North Indian, Thai                        800.0   
2                      Cafe, Mexican, Italian                        800.0   
3                  South Indian, North Indian                        300.0   
4                    North Indian, Rajasthani                        600.0   
...                                       ...                          ...   
51712                             Continental                       1500.0   
51713                             Finger Food                        600.0   
51714                             Finger Food                       2000.0   
51715                             Finger Food                       2500.0   
51716  Finger Food, North Indian, Continental                       1500.0   

                                            reviews_list menu_item  \
0      [('Rated 4.0', 'RATED\n  A beautiful place to ...        []   
1      [('Rated 4.0', 'RATED\n  Had been here for din...        []   
2      [('Rated 3.0', "RATED\n  Ambience is not that ...        []   
3      [('Rated 4.0', "RATED\n  Great food and proper...        []   
4      [('Rated 4.0', 'RATED\n  Very good restaurant ...        []   
...                                                  ...       ...   
51712  [('Rated 5.0', "RATED\n  Food and service are ...        []   
51713                                                 []        []   
51714                                                 []        []   
51715  [('Rated 4.0', 'RATED\n  Nice and friendly pla...        []   
51716  [('Rated 5.0', 'RATED\n  Great ambience , look...        []   

      listed_in(type) listed_in(city)  
0              Buffet    Banashankari  
1              Buffet    Banashankari  
2              Buffet    Banashankari  
3              Buffet    Banashankari  
4              Buffet    Banashankari  
...               ...             ...  
51712   Pubs and bars      Whitefield  

In [28]:
type(df['menu_item'][0])

str

In [29]:
df['menu_item'].unique()

array(['[]',
       "['Chocolate Fantasy (Pack Of 5)', 'Pan Cake (Pack Of 6)', 'Gulab Jamun (Pack Of 10)', 'Gulkand Shot (Pack Of 5)', 'Chocolate Decadence (Pack of 2)', 'CheeseCake (Pack Of 2)', 'Red Velvet Slice Cake (Pack of 2)', 'Red Velvet Slice Cake & Cheese Cake (Pack of 2)', 'Red Velvet Slice Cake & Chocolate Decadence Cake (Pack of 2)', 'Hazelnut Brownie (Pack of 2)', 'Moments', 'Red Velvet Cake With Butter Cream Frosting (750 Gm)', 'Red Velvet Slice Cake (Pack of 2)', 'Red Velvet Slice Cake & Cheese Cake (Pack of 2)', 'Red Velvet Slice Cake & Chocolate Decadence Cake (Pack of 2)', 'Red Velvet Slice Cake (Pack of 1)', 'Valentine Red Velvet Jar', 'Valentine Chocolate Jar', 'Valentines Jar Combo', 'Pink Guava 500 ML', 'Oreo Vanilla 500 ML', 'Cookie Crumble 500 ML', 'Chocolate Fantasy', 'Gulkand-E-Bahar', 'Pan Cake', 'Hazelnut Brownie (Pack Of 1)', 'Gulab Jamun (Pack Of 2)', 'Plum Cake', 'Red Velvet Cake With Butter Cream Frosting (750 Gm)', 'Chocolate Mud Cake (700 Gms)', 'Chees

In [30]:
df

online_order  book_table    rate  votes                    location  \
0                 1           1   4.1/5    775                Banashankari   
1                 1           0   4.1/5    787                Banashankari   
2                 1           0   3.8/5    918                Banashankari   
3                 0           0   3.7/5     88                Banashankari   
4                 0           0   3.8/5    166                Basavanagudi   
...             ...         ...     ...    ...                         ...   
51712             0           0  3.6 /5     27                  Whitefield   
51713             0           0     NaN      0                  Whitefield   
51714             0           0     NaN      0                  Whitefield   
51715             0           1  4.3 /5    236  ITPL Main Road, Whitefield   
51716             0           0  3.4 /5     13  ITPL Main Road, Whitefield   

                 rest_type                                         dish_liked  \
0            Casual Dining  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1            Casual Dining  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2      Cafe, Casual Dining  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3              Quick Bites                                        Masala Dosa   
4            Casual Dining                                Panipuri, Gol Gappe   
...                    ...                                                ...   
51712                  Bar                                                NaN   
51713                  Bar                                                NaN   
51714                  Bar                                                NaN   
51715                  Bar                       Cocktails, Pizza, Buttermilk   
51716   Bar, Casual Dining                                                NaN   

                                     cuisines  approx_cost(for two people)  \
0              North Indian, Mughlai, Chinese                        800.0   
1                 Chinese, North Indian, Thai                        800.0   
2                      Cafe, Mexican, Italian                        800.0   
3                  South Indian, North Indian                        300.0   
4                    North Indian, Rajasthani                        600.0   
...                                       ...                          ...   
51712                             Continental                       1500.0   
51713                             Finger Food                        600.0   
51714                             Finger Food                       2000.0   
51715                             Finger Food                       2500.0   
51716  Finger Food, North Indian, Continental                       1500.0   

                                            reviews_list menu_item  \
0      [('Rated 4.0', 'RATED\n  A beautiful place to ...        []   
1      [('Rated 4.0', 'RATED\n  Had been here for din...        []   
2      [('Rated 3.0', "RATED\n  Ambience is not that ...        []   
3      [('Rated 4.0', "RATED\n  Great food and proper...        []   
4      [('Rated 4.0', 'RATED\n  Very good restaurant ...        []   
...                                                  ...       ...   
51712  [('Rated 5.0', "RATED\n  Food and service are ...        []   
51713                                                 []        []   
51714                                                 []        []   
51715  [('Rated 4.0', 'RATED\n  Nice and friendly pla...        []   
51716  [('Rated 5.0', 'RATED\n  Great ambience , look...        []   

      listed_in(type) listed_in(city)  
0              Buffet    Banashankari  
1              Buffet    Banashankari  
2              Buffet    Banashankari  
3              Buffet    Banashankari  
4              Buffet    Banashankari  
...               ...             ...  
51712   Pubs and bars      Whitefield  

In [31]:
type(df["rate"])

pandas.core.series.Series

In [32]:
# def remove1(s):
#     ans = 0
#     for i in s:
#         if i =='/':
#             return ans
            
#         ans *= 10
#         ans += int(i)
#         return ans
    
        

In [33]:
#for i in range (len(df['rate'])):
#    df.at[i,'rate']=remove1(df.at[i,'rate'])
    

In [34]:
df['votes'].isnull().sum()

0

In [35]:
df['location'].isnull().sum()

21

In [36]:
df['location'].unique()

array(['Banashankari', 'Basavanagudi', 'Mysore Road', 'Jayanagar',
       'Kumaraswamy Layout', 'Rajarajeshwari Nagar', 'Vijay Nagar',
       'Uttarahalli', 'JP Nagar', 'South Bangalore', 'City Market',
       'Nagarbhavi', 'Bannerghatta Road', 'BTM', 'Kanakapura Road',
       'Bommanahalli', nan, 'CV Raman Nagar', 'Electronic City', 'HSR',
       'Marathahalli', 'Sarjapur Road', 'Wilson Garden', 'Shanti Nagar',
       'Koramangala 5th Block', 'Koramangala 8th Block', 'Richmond Road',
       'Koramangala 7th Block', 'Jalahalli', 'Koramangala 4th Block',
       'Bellandur', 'Whitefield', 'East Bangalore', 'Old Airport Road',
       'Indiranagar', 'Koramangala 1st Block', 'Frazer Town', 'RT Nagar',
       'MG Road', 'Brigade Road', 'Lavelle Road', 'Church Street',
       'Ulsoor', 'Residency Road', 'Shivajinagar', 'Infantry Road',
       'St. Marks Road', 'Cunningham Road', 'Race Course Road',
       'Commercial Street', 'Vasanth Nagar', 'HBR Layout', 'Domlur',
       'Ejipura', 'Jeevan 

In [37]:
df['location'] = df['location'].dropna()

In [38]:
df['location'].isnull().sum()

21

In [39]:
df['location'].isnull().iloc()

In [40]:
df[df['location'] == True].index.tolist()


[]

In [41]:
df.index[df['location']=='nan'].tolist()


[]

In [151]:
df['location'].isnull()

0        False
1        False
2        False
3        False
4        False
         ...  
51705    False
51707    False
51708    False
51711    False
51715    False
Name: location, Length: 23392, dtype: bool

### BUILDING A FUNCTION TO DROP ALL THE NULL VALUES IN THE 'location' COLUMN

In [43]:
arr = df['location'].isnull()
for i in range(arr.size):
    if(arr[i] == True):
        df.drop([i] , axis = 0,inplace = True)

In [44]:
df['cuisines'].isnull().sum()

24

In [45]:
df['cuisines'].unique()

array(['North Indian, Mughlai, Chinese', 'Chinese, North Indian, Thai',
       'Cafe, Mexican, Italian', ...,
       'North Indian, Street Food, Biryani', 'Chinese, Mughlai',
       'North Indian, Chinese, Arabian, Momos'], dtype=object)

In [46]:
arr1 = df['cuisines'].unique()
print(arr1.size)

2724


In [47]:
arr2 = df['dish_liked'].unique()
print(arr2.size)

5272


In [48]:
df.dropna(how='any', inplace=True)


In [142]:
df.shape

(23392, 13)

### ON THE 'rate' COLUMN WE'RE EXTRACTING THE NUMERICAL VALUE WHICH IS  THE ACTUAL RATING 

In [50]:
df['rate'] = df['rate'].apply(lambda x: x.replace('/5',''))

In [53]:
df['rate'] = df['rate'].astype(float)

/var/folders/jm/fzpx9nmx3gl9k9bfkhp1cy100000gn/T/ipykernel_21032/2854580240.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['rate'] = df['rate'].astype(float)


In [52]:
df = df.loc[df.rate != 'NEW']

In [143]:
df

online_order  book_table  rate  votes  location  rest_type  \
0                 1           1   4.1    775         1         20   
1                 1           0   4.1    787         1         20   
2                 1           0   3.8    918         1         16   
3                 0           0   3.7     88         1         62   
4                 0           0   3.8    166         4         20   
...             ...         ...   ...    ...       ...        ...   
51705             1           1   3.8    128        84          5   
51707             0           0   3.9     77        84         38   
51708             0           0   2.8    161        84         21   
51711             0           0   2.5     81        84         21   
51715             0           1   4.3    236        25          4   

                                              dish_liked  cuisines  \
0      Pasta, Lunch Buffet, Masala Papad, Paneer Laja...      1391   
1      Momos, Lunch Buffet, Chocolate Nirvana, Thai G...       596   
2      Churros, Cannelloni, Minestrone Soup, Hot Choc...       486   
3                                            Masala Dosa      1592   
4                                    Panipuri, Gol Gappe      1411   
...                                                  ...       ...   
51705  Beer, Chicken Guntur, Paneer Tikka, Fish, Nood...      1323   
51707                                   Rooftop Ambience       882   
51708  Salads, Coffee, Breakfast Buffet, Halwa, Chick...       541   
51711                               Biryani, Andhra Meal        80   
51715                       Cocktails, Pizza, Buttermilk       882   

       approx_cost(for two people)  \
0                            800.0   
1                            800.0   
2                            800.0   
3                            300.0   
4                            600.0   
...                            ...   
51705                       1200.0   
51707                       2000.0   
51708                       1200.0   
51711                        800.0   
51715                       2500.0   

                                            reviews_list  menu_item  \
0      [('Rated 4.0', 'RATED\n  A beautiful place to ...       5047   
1      [('Rated 4.0', 'RATED\n  Had been here for din...       5047   
2      [('Rated 3.0', "RATED\n  Ambience is not that ...       5047   
3      [('Rated 4.0', "RATED\n  Great food and proper...       5047   
4      [('Rated 4.0', 'RATED\n  Very good restaurant ...       5047   
...                                                  ...        ...   
51705  [('Rated 3.0', "RATED\n  Nice place to hangout...       5047   
51707  [('Rated 4.0', 'RATED\n  Went there post dinne...       5047   
51708  [('Rated 3.0', 'RATED\n  Place is good not tha...       5047   
51711  [('Rated 4.0', 'RATED\n  A fine place to chill...       5047   
51715  [('Rated 4.0', 'RATED\n  Nice and friendly pla...       5047   

      listed_in(type) listed_in(city)  
0              Buffet    Banashankari  
1              Buffet    Banashankari  
2              Buffet    Banashankari  
3              Buffet    Banashankari  
4              Buffet    Banashankari  
...               ...             ...  
51705   Pubs and bars      Whitefield  
51707   Pubs and bars      Whitefield  
51708   Pubs and bars      Whitefield  
51711   Pubs and bars      Whitefield  
51715   Pubs and bars      Whitefield  

[23392 rows x 13 columns]

### WITH THE HELP OF LabelEncoder  WE'RE JUST ASSIGNING THE NUMERICAL CODES FOR THE DIFFERENT VALUES IN THE COLUMNS SO THAT MODEL IS ABLE TO FEED ON THESE COLUMNS

In [55]:
from sklearn.preprocessing import LabelEncoder
la = LabelEncoder() 

In [56]:
df.location = la.fit_transform(df.location)
df.menu_item = la.fit_transform(df.menu_item)
df.cuisines = la.fit_transform(df.cuisines)
df.rest_type = la.fit_transform(df.rest_type)

/var/folders/jm/fzpx9nmx3gl9k9bfkhp1cy100000gn/T/ipykernel_21032/561758820.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.location = la.fit_transform(df.location)
/var/folders/jm/fzpx9nmx3gl9k9bfkhp1cy100000gn/T/ipykernel_21032/561758820.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.menu_item = la.fit_transform(df.menu_item)
/var/folders/jm/fzpx9nmx3gl9k9bfkhp1cy100000gn/T/ipykernel_21032/561758820.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFr

In [146]:
df

online_order  book_table  rate  votes  location  rest_type  \
0                 1           1   4.1    775         1         20   
1                 1           0   4.1    787         1         20   
2                 1           0   3.8    918         1         16   
3                 0           0   3.7     88         1         62   
4                 0           0   3.8    166         4         20   
...             ...         ...   ...    ...       ...        ...   
51705             1           1   3.8    128        84          5   
51707             0           0   3.9     77        84         38   
51708             0           0   2.8    161        84         21   
51711             0           0   2.5     81        84         21   
51715             0           1   4.3    236        25          4   

                                              dish_liked  cuisines  \
0      Pasta, Lunch Buffet, Masala Papad, Paneer Laja...      1391   
1      Momos, Lunch Buffet, Chocolate Nirvana, Thai G...       596   
2      Churros, Cannelloni, Minestrone Soup, Hot Choc...       486   
3                                            Masala Dosa      1592   
4                                    Panipuri, Gol Gappe      1411   
...                                                  ...       ...   
51705  Beer, Chicken Guntur, Paneer Tikka, Fish, Nood...      1323   
51707                                   Rooftop Ambience       882   
51708  Salads, Coffee, Breakfast Buffet, Halwa, Chick...       541   
51711                               Biryani, Andhra Meal        80   
51715                       Cocktails, Pizza, Buttermilk       882   

       approx_cost(for two people)  \
0                            800.0   
1                            800.0   
2                            800.0   
3                            300.0   
4                            600.0   
...                            ...   
51705                       1200.0   
51707                       2000.0   
51708                       1200.0   
51711                        800.0   
51715                       2500.0   

                                            reviews_list  menu_item  \
0      [('Rated 4.0', 'RATED\n  A beautiful place to ...       5047   
1      [('Rated 4.0', 'RATED\n  Had been here for din...       5047   
2      [('Rated 3.0', "RATED\n  Ambience is not that ...       5047   
3      [('Rated 4.0', "RATED\n  Great food and proper...       5047   
4      [('Rated 4.0', 'RATED\n  Very good restaurant ...       5047   
...                                                  ...        ...   
51705  [('Rated 3.0', "RATED\n  Nice place to hangout...       5047   
51707  [('Rated 4.0', 'RATED\n  Went there post dinne...       5047   
51708  [('Rated 3.0', 'RATED\n  Place is good not tha...       5047   
51711  [('Rated 4.0', 'RATED\n  A fine place to chill...       5047   
51715  [('Rated 4.0', 'RATED\n  Nice and friendly pla...       5047   

      listed_in(type) listed_in(city)  
0              Buffet    Banashankari  
1              Buffet    Banashankari  
2              Buffet    Banashankari  
3              Buffet    Banashankari  
4              Buffet    Banashankari  
...               ...             ...  
51705   Pubs and bars      Whitefield  
51707   Pubs and bars      Whitefield  
51708   Pubs and bars      Whitefield  
51711   Pubs and bars      Whitefield  
51715   Pubs and bars      Whitefield  

[23392 rows x 13 columns]

### EXTRACTING THE COLUMNS THROUGH WHICH MODEL WILL LEARN TO PREDICT OR WE CAN SAY WITH THE HELP OF THESE MODEL WILL MAKE PREDICTIONS (THEY ARE THE SIGNIFICANT FACTORS)

In [145]:
x = df.iloc[:,[0,1,3,4,5,7,8,10]]
x.head()

online_order  book_table  votes  location  rest_type  cuisines  \
0             1           1    775         1         20      1391   
1             1           0    787         1         20       596   
2             1           0    918         1         16       486   
3             0           0     88         1         62      1592   
4             0           0    166         4         20      1411   

   approx_cost(for two people)  menu_item  
0                        800.0       5047  
1                        800.0       5047  
2                        800.0       5047  
3                        300.0       5047  
4                        600.0       5047

###  EXTRACTING THE COLUMN WHICH IS TO BE PREDICTED BY THE MODEL

In [64]:
y = df['rate']
y

0        4.1
1        4.1
2        3.8
3        3.7
4        3.8
        ... 
51705    3.8
51707    3.9
51708    2.8
51711    2.5
51715    4.3
Name: rate, Length: 23392, dtype: float64

# MODEL BUILDING


In [67]:
from sklearn.model_selection import train_test_split

In [70]:
#help(train_test_split)

In [99]:
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.33, random_state=42)

### APPLYING LINEAR REGRESSION

In [73]:
from sklearn.linear_model import LinearRegression

In [75]:
Li = LinearRegression()

In [81]:
Li.fit(X_train,y_train)

LinearRegression()

In [109]:
Li_prediction_result = Li.predict(X_test)
Li_prediction_result

array([4.15803337, 4.0139357 , 3.66361409, ..., 4.20271065, 3.73426747,
       4.3086058 ])

### PREDICTING THE ERROR OF LINEAR REGRESSION MODEL

In [88]:
from sklearn.metrics import mean_squared_error,mean_absolute_error

In [110]:
mean_absolute_error(y_test,Li_prediction_result)

0.2648613959304658

In [111]:
mean_squared_error(y_test,Li_prediction_result)

0.14160991686635838

In [112]:
from sklearn.metrics import r2_score

In [113]:
r2_score(y_test,Li_prediction_result)

0.219963661359864

### APPLYING RANDOM FOREST REGRESSOR

In [101]:
from sklearn.ensemble import RandomForestRegressor

In [103]:
rf_model = RandomForestRegressor()

In [152]:
rf_model.fit(X_train,y_train)

RandomForestRegressor()

### PREDICTING THE ERROR OF RANDOM FOREST REGRESSOR MODEL

In [116]:
rf_prediction_result = rf_model.predict(X_test)

In [118]:
mean_absolute_error(y_test,rf_prediction_result)

0.05890816424436353

In [120]:
mean_squared_error(y_test,rf_prediction_result)

0.018489335588347814

In [122]:
r2_score(y_test,rf_prediction_result)

0.8981543527785952

### APPLYING EXTRA TREE REGRESSOR

In [124]:
from sklearn.ensemble import ExtraTreesRegressor

In [126]:
ex_model = ExtraTreesRegressor()

In [158]:
ex_model.fit(X_train,y_train)

ExtraTreesRegressor()

### PREDICTING THE ERROR OF EXTRA TREE REGRESSOR MODEL

In [134]:
ex_prediction_result = ex_model.predict(X_test)
ex_prediction_result

array([4.3  , 3.8  , 4.   , ..., 4.201, 4.   , 4.1  ])

In [154]:
r2_score(y_test,ex_prediction_result)

0.9215750127190919

In [156]:
mean_squared_error(y_test,ex_prediction_result)

0.014237485331075285

In [159]:
mean_absolute_error(y_test,ex_prediction_result)

0.03664347181100749

# IN CONCLUSION THE ACCURACY OF THE MODEL IS 92%(USING EXTRA TREE REGRESSOR) AND ERROR IS 0.014 FOR RMSE AND 0.036 FOR MAE . 